Teach_Breakout.py

```diff
@@ -0,0 +1,344 @@
+                one_hot_actions[i, j] = 1
+
+    # Fit the keras model, using the mask
+    model.fit([start_states, one_hot_actions], one_hot_actions * Q_values[:, None],
+        epochs=1, batch_size=len(start_states), verbose=0) # Should it be epochs=1??
+
+
+def make_n_fit_batch(model, target_model, gamma, iteration, mem_size, mem_frames, mem_actions, mem_rewards, mem_is_terminal, number_in_batch):
+    """Make a batch then use it to train the model"""
+
+    if iteration < mem_size:
+        # In this case, sample 'number_in_batch' numbers from 4 to 'iterations'
+        # (We use 4 so that there are 5 frames minimum, so we can still make 2 states (because each state is the frame
+        # plus the previous 3 frames))
+        indices_chosen = random.sample(range(4, iteration), number_in_batch)
+    else:
+        # Now the memory is full, so we can take any elements from it
+        indices_chosen = random.sample(range(0, mem_size), number_in_batch)
+
+    # Initialise the batches
+    start_states = np.zeros((number_in_batch, 105, 80, 4))  # (Array instead of list)
+    next_states = np.zeros((number_in_batch, 105, 80, 4))  # (Array instead of list)
+    actions = np.zeros((number_in_batch))  # (Array instead of list)
+
+    rewards = list()  # List
+    is_terminals = list()  # List
+
+    for i in range(len(indices_chosen)): # Probably more efficient way to do this: "for index in ?(indices_chosen)?"
+        index = indices_chosen[i] # index corresponds to the iterations
+        start_states[i, :, :, :] = make_state(mem_frames, index - 1)
+        next_states[i, :, :, :] = make_state(mem_frames, index) # State given by index was arrived at by taking action given by index and reward received is give by index.
+                                                                # In contrast, index-1 labels the previous state, before the action was taken
+        actions[i] = mem_actions.recall(index)
+
+        rewards.append(mem_rewards.recall(index))
+        is_terminals.append(mem_is_terminal.recall(index))
+
+    # We should now have a full batch, which the DNN can train on
+    fit_batch_target(model, target_model, gamma, start_states, actions, rewards, next_states, is_terminals, number_in_batch)
+
+
+def q_iteration(env, model, target_model, iteration, mem_frames, mem_actions, mem_rewards, mem_is_terminal, mem_size):
+    """
+    Do one iteration of acting then learning
+    """
+
+    # Choose epsilon based on the iteration
+    epsilon = get_epsilon_for_iteration(iteration)
+
+    start_state = make_state(mem_frames, iteration)
+    # Choose the action:
+    if random.random() < epsilon:
+        action = env.action_space.sample()
+    else:
+        action = choose_best_action(model, start_state)
+    # Play one game iteration (note: according to a paper, you should actually play 4 times here)
+    new_frame, reward, is_terminal, _ = env.step(action)
+
+    reward = transform_reward(reward)  # Make reward just +1 or -1
```

Comment:
There's some code duplication between lines 218-225 and lines 301-308. Can you factor out common code (preprocessing, adding to `mem_frames`, ...)?